File: problem_17/problem_17.py
import time

def problem_17():
    ones = [len(n) for n in [
        "one","two","three","four","five",
        "six","seven","eight","nine",
    ]]
    teens = [len(n) for n in [
        "eleven","twelve","thirteen","fourteen","fifteen",
        "sixteen","seventeen","eighteen","nineteen",
    ]]
    tens = [len(n) for n in [
        "ten","twenty","thirty","forty","fifty",
        "sixty","seventy","eighty","ninety",
    ]]
    hundreds = [7+n for n in ones]

    letters = 0

    letters += sum(ones) + sum(teens) + sum(tens) + sum(hundreds)
    letters += sum([ten + one for one in ones for ten in tens[1::]])
    letters += sum([h + 3 + one for one in ones for h in hundreds])
    letters += sum([h + 3 + t for t in teens for h in hundreds])
    letters += sum([h + 3 + t for t in tens for h in hundreds])
    letters += sum([h + 3 + t + one for one in ones for t in tens[1::] for h in hundreds])
    letters += 11 # one thousand
    
    return letters

start = time.time()
solution = problem_17()
print(f"{solution} found in {time.time() - start} seconds.")
# 21124 found in 0.00010371208190917969 seconds.

File: problem_17/problem_17.js
const now = require("performance-now");

const problem17 = () => {

    const sum = arr => arr.reduce((a,b)=>(a+b));

    const ones = [
        "one","two","three",
        "four","five","six",
        "seven","eight","nine"
    ].map(e => e.length);
    
    const teens = [
        "eleven","twelve","thirteen",
        "fourteen","fifteen","sixteen",
        "seventeen","eighteen","nineteen"
    ].map(e => e.length);
    
    const tens = [
        "ten","twenty","thirty",
        "forty","fifty","sixty",
        "seventy","eighty","ninety"
    ].map(e => e.length);
    
    const hundreds = [
        "onehundred","twohundred","threehundred",
        "fourhundred","fivehundred","sixhundred",
        "sevenhundred","eighthundred","ninehundred"
    ].map(e => e.length);
    
    let letters = 0;
    
    letters += sum(ones) + sum(teens) + sum(tens) + sum(hundreds)

    for (const ten of tens.slice(1,)){
        for (const one of ones){
            letters += ten + one;
        }
    }
    for (const hundred of hundreds){
        for (const one of ones){
            letters += hundred + 3 + one;
        }
        letters += hundred + 3 + tens[0];
        for (const teen of teens){
            letters += hundred + 3 + teen
        }
        for (const ten of tens.slice(1,)){
            letters += hundred + 3 + ten
            for (const one of ones){
                letters += hundred + 3 + ten + one;
            }
        }
    }
    letters += 11; // onethousand
    return letters;
}

let start = now()
let solution = problem17()
console.log(`${solution} found in ${(now()-start) / 1000} seconds.`)
// 21124 found in 0.0003209489999999953 seconds.